File: 프로그래머스/월간 코드 챌린지 시즌3/빛의 경로 사이클/빛의 경로 사이클.py
def solution(grid):
    # visit[i][j][k]: (i, j)에서 k방향으로 간 적이 있는지
    visit = [[[False for i in range(4)] 
              for i in range(len(grid[0]))] for i in range(len(grid))]
    
    # dir: 0(상)1(우)2(하)3(좌)
    answer = []
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            for k in range(4):
                if not visit[i][j][k]:
                    visit[i][j][k] = True
                    cnt=0; x,y=i,j; dir=k

                    while True:
                        cnt += 1

                        # 좌표 움직이기
                        if dir == 0: x -= 1
                        elif dir == 1: y += 1
                        elif dir == 2: x += 1
                        else: y -= 1

                        # 범위 벗어났는지 체크
                        if x < 0: x = len(grid)-1
                        elif x >= len(grid): x = 0

                        if y < 0: y = len(grid[0])-1
                        elif y >= len(grid[0]): y = 0

                        # 다음 방향 계산
                        if grid[x][y] == 'L':
                            dir -= 1
                            if dir < 0: dir += 4
                        elif grid[x][y] == 'R':
                            dir += 1
                            dir %= 4

                        # 방문 체크
                        if visit[x][y][dir]: break
                        else: visit[x][y][dir] = True
                    
                    answer.append(cnt)

    return sorted(answer)

File: 프로그래머스/월간 코드 챌린지 시즌3/빛의 경로 사이클/빛의 경로 사이클.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> solution(vector<string> grid) {
    // visit[i][j][k]: (i,j)에서 k방향으로 간 적이 있는지
    vector<vector<vector<bool>>> visit;
    for(int i=0; i<grid.size(); i++){
        vector<vector<bool>> v;
        for(int j=0; j<grid[0].size(); j++){
            vector<bool> vv(4, false);
            v.push_back(vv);
        }
        visit.push_back(v);
    }
    
    // dir: 0(상)1(우)2(하)3(좌)
    vector<int> answer;
    for(int i=0; i<grid.size(); i++){
        for(int j=0; j<grid[0].size(); j++){
            for(int k=0; k<4; k++){
                if(!visit[i][j][k]){
                    visit[i][j][k] = true;
                    int cnt=0, x=i, y=j, dir=k;
                    
                    while(1){
                        cnt++;
                        
                        // 좌표 움직이기
                        switch(dir){
                            case 0: x--; break;
                            case 1: y++; break;
                            case 2: x++; break;
                            case 3: y--; break;
                        }
                        
                        // 범위 벗어났는지 체크
                        if(x < 0) x = grid.size()-1;
                        else if(x >= grid.size()) x = 0;
                        
                        if(y < 0) y = grid[0].size()-1;
                        else if(y >= grid[0].size()) y = 0;
                        
                        // 다음 방향 계산
                        switch(grid[x][y]){
                            case 'L': dir--; if(dir < 0) dir += 4;
                                break;
                            case 'R': dir++; dir%=4;
                                break;
                        }
                        
                        // 방문 체크
                        if(visit[x][y][dir]) break;
                        else visit[x][y][dir] = true;
                    }
                    answer.push_back(cnt);
                }
            }
        }
    }
    
    sort(answer.begin(), answer.end());
    return answer;
}
